### PR TITLE
Implement logging for LibSourcify

### PR DESCRIPTION
**Pull request by marcocastignoli**
* Created the simplest logging library inside LibSourcify (with no deps)
* change the default log level with `setLibSourcifyLoggerLevel`
* override the whole logging function with `setLibSourcifyLogger` implementing `ILibSourcifyLogger`

use the logger inside lib sourcify with `logError`, `logWarn`, `logInfo`, `logDebug`

@kuzdogan we need to decide what to log inside lib-sourcify

### CODE REVIEW

File: packages/lib-sourcify/src/lib/CheckedContract.ts [hunk 1]
```diff
@@ -338,29 +340,29 @@ export async function performFetch(
   hash?: string,
   fileName?: string
 ): Promise<string | null> {
-  console.log(`Fetching the file ${fileName} from ${url}...`);
+  logInfo(`Fetching the file ${fileName} from ${url}...`);
```
Comment (marcocastignoli) [inline packages/lib-sourcify/src/lib/CheckedContract.ts:343]: This is an example of how to use the logger inside lib-sourcify

File: src/server/server.ts [hunk 2]
```diff
@@ -28,11 +28,39 @@ import { resolveRefs } from "json-refs";
...
+          logger.info(msg);
+          break;
+        case 4:
+          logger.debug(msg);
+          break;
+      }
+    }
+  },
+});
+
```
Comment (marcocastignoli) [inline src/server/server.ts:63]: This is how you override the default lib-sourcify's logger

File: packages/lib-sourcify/src/lib/logger.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
+export interface ILogger {
+  logLevel: number;
+  log: (level: number, message: string) => void;
+  setLevel: (level: number) => void;
+}
+
+// Default logger behavior
+export const DefaultLogger: ILogger = {
+  logLevel: 1,
```
Comment (kuzdogan) [inline packages/lib-sourcify/src/lib/logger.ts:9]: Isn't default 3?

**Comment by kuzdogan** [comment]
- [ ] Add how to override the logger in `lib-sourcify`'s `README`

File: packages/lib-sourcify/src/lib/logger.ts (hunk 3 above)
Comment (marcocastignoli) [reply packages/lib-sourcify/src/lib/logger.ts:9]: From a quick search most libraries implement the warn level

File: src/server/server.ts [hunk 4]
```diff
@@ -28,11 +28,39 @@ import { resolveRefs } from "json-refs";
...
+import { setLibSourcifyLogger } from "@ethereum-sourcify/lib-sourcify";
 // eslint-disable-next-line @typescript-eslint/no-var-requires
 const fileUpload = require("express-fileupload");
 
 const MemoryStore = createMemoryStore(session);
 
+// here we override the standard LibSourcify's Logger with a custom one
+setLibSourcifyLogger({
+  // No need to set again the logger level because it's set here
+  logLevel: 4,
```
Comment (kuzdogan) [inline src/server/server.ts:40]:
```suggestion
  logLevel: process.env.NODE_ENV === "production" ? 3 : 4,
```

File: packages/lib-sourcify/src/lib/logger.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
+export interface ILogger {
+  logLevel: number;
+  log: (level: number, message: string) => void;
+  setLevel: (level: number) => void;
+}
+
+// Default logger behavior
+export const DefaultLogger: ILogger = {
+  logLevel: 2,
```
Comment (kuzdogan) [inline packages/lib-sourcify/src/lib/logger.ts:9]: We should log info level, no?
```suggestion
  logLevel: 3,
```

**Review by kuzdogan — CHANGES_REQUESTED** [review]
Nice! 

There needs to be more work on what to log and on which level to log, but we can work on that later as we see the needs. 

Just asking for small changes and clarifications.

File: packages/lib-sourcify/src/lib/logger.ts (hunk 5 above)
Comment (marcocastignoli) [reply packages/lib-sourcify/src/lib/logger.ts:9]: I don't like library that by default they log info, I see most of them simply log warn and err

File: packages/lib-sourcify/README.md [hunk 6]
```diff
@@ -104,3 +104,49 @@ const match = await verifyCreate2(
...
 ```
+
+## Logging
+
+`lib-sourcify` has a basic logging system.
+
+You can specify the log level using the `setLibSourcifyLoggerLevel(level)` where:
+
+- `0` is nothing
+- `1` is errors _[default]_
```
[the remark this answers is not in the document]
Comment (marcocastignoli) [reply packages/lib-sourcify/README.md:115]: I forgot to update it here

**Pull request merged**